source/simulator/core.cc

```diff
@@ -2147,6 +2167,286 @@ namespace aspect
+                                                                  parameters.linear_stokes_solver_tolerance,
+                                                                  stokes_residual,
+                                                                  residual,
+                                                                  residual_old);
+
+                      pcout << "   The linear solver tolerance is set to " << parameters.linear_stokes_solver_tolerance << std::endl;
+                    }
+                }
+
+              build_stokes_preconditioner();
+
+              stokes_residual = solve_stokes();
+
+              velocity_residual = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+              pressure_residual = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+              residual = std::sqrt(velocity_residual * velocity_residual + pressure_residual * pressure_residual);
+
+              /**
+               * We may need to do a line search if the solution update doesn't decrease the norm of the rhs enough.
+               * This is done by adding the solution update to the current linearization point and then assembling
+               * the Newton right hand side. If the Newton residual has decreased enough by using the this update,
+               * then we continue, otherwise we reset the current linearization point with the help of the backup and
+               * add each iteration an increasingly smaller solution update until the decreasing residual condition
+               * is met, or the line search iteration limit is reached.
+               */
+              LinearAlgebra::BlockVector backup_linearization_point = current_linearization_point;
+
+              double test_residual = 0;
+              double test_velocity_residual = 0;
+              double test_pressure_residual = 0;
+              double lambda = 1;
+              double alpha = 1e-4;
+              unsigned int line_search_iteration = 0;
+
+              /**
+               * Do the loop for the line search. Even when we
+               * don't do a line search we go into this loop
+               */
+              do
+                {
+                  current_linearization_point = backup_linearization_point;
+
+                  LinearAlgebra::BlockVector search_direction = solution;
+
+                  search_direction *= lambda;
+
+
+                  current_linearization_point.block(introspection.block_indices.pressure) += search_direction.block(introspection.block_indices.pressure);
+                  current_linearization_point.block(introspection.block_indices.velocities) += search_direction.block(introspection.block_indices.velocities);
+
+                  assemble_newton_stokes_matrix = rebuild_stokes_preconditioner = false;
+                  rebuild_stokes_matrix = true;
+
+                  assemble_stokes_system();
```

Comment:
`assemble_newton_stokes_matrix` determines whether only the newton rhs is build or the whole newton system. 

`rebuild_stokes_matrix` determines whether information for rebuilding the Stokes matrix needs to be (re)computed. For the Newton solver this variable has to be true all the time because even the rhs needs this info.